server/sonar-process/src/main/java/org/sonar/process/logging/Log4JPropertiesBuilder.java

```diff
@@ -0,0 +1,279 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process.logging;
+
+import ch.qos.logback.classic.Level;
+import java.io.File;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.process.MessageException;
+import org.sonar.process.Props;
+
+import static java.lang.String.format;
+import static java.lang.String.valueOf;
+
+public class Log4JPropertiesBuilder extends AbstractLogHelper {
+  private static final String ROOT_LOGGER_NAME = "rootLogger";
+
+  private final Properties log4j2Properties = new Properties();
+  private final Props props;
+
+  public Log4JPropertiesBuilder(Props props) {
+    super("%logger{1.}");
+    this.props = Objects.requireNonNull(props, "Props can't be null");
+    internalLogLevel(Level.ERROR);
+  }
+
+  @Override
+  public String getRootLoggerName() {
+    return ROOT_LOGGER_NAME;
+  }
+
+  public Properties get() {
+    Properties res = new Properties();
+    res.putAll(log4j2Properties);
+    return res;
+  }
+
+  public void internalLogLevel(Level level) {
+    putProperty("status", level.toString());
+  }
+
+  private void putProperty(String key, String value) {
+    log4j2Properties.put(key, value);
+  }
+
+  /**
+   * Make log4j2 configuration for a process to push all its logs to a log file.
+   * <p>
+   * <ul>
+   * <li>the file's name will use the prefix defined in {@link RootLoggerConfig#getProcessId()#getLogFilenamePrefix()}.</li>
+   * <li>the file will follow the rotation policy defined in property {@link #ROLLING_POLICY_PROPERTY} and
+   * the max number of files defined in property {@link #MAX_FILES_PROPERTY}</li>
+   * <li>the logs will follow the specified log pattern</li>
+   * </ul>
+   * </p>
+   *
+   * @see #buildLogPattern(RootLoggerConfig)
+   */
+  public void configureGlobalFileLog(RootLoggerConfig config, File logDir, String logPattern) {
+    String appenderRef = writeFileAppender(config, logDir, logPattern);
+
+    putProperty(ROOT_LOGGER_NAME + ".appenderRef." + appenderRef + ".ref", appenderRef);
+  }
+
+  private String writeFileAppender(RootLoggerConfig config, File logDir, String logPattern) {
+    String appenderName = "file_" + config.getProcessId().getLogFilenamePrefix();
+    RollingPolicy rollingPolicy = createRollingPolicy(logDir, config.getProcessId().getLogFilenamePrefix());
+    Appender appender = new Appender(appenderName, rollingPolicy, logPattern);
+    appender.writeAppenderProperties();
+    return appender.getAppenderRef();
+  }
+
+  private RollingPolicy createRollingPolicy(File logDir, String filenamePrefix) {
+    String rollingPolicy = props.value(ROLLING_POLICY_PROPERTY, "time:yyyy-MM-dd");
+    int maxFiles = props.valueAsInt(MAX_FILES_PROPERTY, 7);
+
+    if (rollingPolicy.startsWith("time:")) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "rollingPolicy" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)